services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,572 @@
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(resourceVersion);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> processDeleteTenant(tenantId, resourceVersion, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<Result<Void>> processDeleteTenant(final String tenantId, final Optional<String> resourceVersion,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(resourceVersion);
+
+        final JsonObject tenantDeleteQuery = resourceVersion
+                .map(version -> new MongoDbDocumentBuilder().withVersion(version))
+                .orElse(new MongoDbDocumentBuilder())
+                .withTenantId(tenantId)
+                .document();
+
+        final Promise<MongoClientDeleteResult> deleteTenantPromise = Promise.promise();
+        mongoClient.removeDocument(config.getCollectionName(), tenantDeleteQuery, deleteTenantPromise);
+        return deleteTenantPromise.future()
+                .compose(tenantDeleteResult -> {
+                    if (tenantDeleteResult.getRemovedCount() == 0) {
+                        return getTenantNotFoundOrVersionMismatch(tenantId, resourceVersion);
+                    } else {
+                        span.log(String.format("successfully deleted tenant [%s]", tenantId));
+                        return Future.succeededFuture(Result.from(HttpURLConnection.HTTP_NO_CONTENT));
+                    }
+                });
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final String tenantId) {
+        Objects.requireNonNull(tenantId);
+
+        return get(tenantId, null);
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+
+        return readTenant(tenantId, span)
+                .compose(successTenantRead -> {
+                    if (successTenantRead.getStatus() != HttpURLConnection.HTTP_OK) {
+                        TracingHelper.logError(span, String.format("tenant [%s] not found.", tenantId));
+                        return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_NOT_FOUND));
+                    }
+                    return Future.succeededFuture(TenantResult.from(
+                            HttpURLConnection.HTTP_OK,
+                            DeviceRegistryUtils
+                                    .convertTenant(tenantId, successTenantRead.getPayload(), true),
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())));
+                });
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn) {
+        Objects.requireNonNull(subjectDn);
+
+        return get(subjectDn, null);
```

Comment:
Please use `NoopSpan.INSTANCE` instead of `null`.

Reply:
done